# Customer Churn Prediction
## Notebook 5 of 8 — Model Building (Baseline)

This project predicts which telecom customers are likely to **churn** (cancel their service) so the business can reach them with retention offers *before* they leave. It walks through the full data-science lifecycle — exploration, cleaning, EDA, feature engineering, modelling, evaluation, and a tuned final model.

**Business problem:** Winning a new customer costs far more than keeping an existing one. This telecom loses roughly **27% of its customers**, and the leadership team wants a reliable, data-driven way to flag at-risk customers early enough to act.

**Tools & techniques:** Python · pandas · NumPy · Matplotlib · Seaborn · scikit-learn · XGBoost · SMOTE (imbalanced-learn) · joblib

> **This notebook:** we split the data and train a Logistic Regression baseline. A simple, honest baseline is the yardstick every later model must beat.

**Author:** La Yaung Linn Lett  &nbsp;·&nbsp;  **Last updated:** June 2026

---

## 1. Load the model-ready data and split features/target

**What:** Load the encoded matrix from Notebook 04 and separate the predictors (`X`) from the target (`y`).

**Why:** Models learn the relationship between `X` (customer attributes) and `y` (whether they churned).

In [1]:
# Standard library
import warnings

# Third-party
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

warnings.filterwarnings("ignore", category=UserWarning)

df = pd.read_csv("../data/processed/model_ready.csv")

# X = all predictors, y = the churn target
X = df.drop(columns=["Churn"])
y = df["Churn"]
print(f"Features: {X.shape}, Target: {y.shape}")

Features: (7043, 29), Target: (7043,)


## 2. Train/test split

**What:** Hold back 20% of customers as an unseen test set.

**Why:** A model must be judged on data it has never seen, otherwise we are just measuring memorisation. `random_state=42` makes the split reproducible.

In [2]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42)

print(f"Train: {X_train.shape[0]} customers")
print(f"Test:  {X_test.shape[0]} customers")

Train: 5634 customers
Test:  1409 customers


**5,634 customers** for training and **1,409** held back for testing — a healthy split for a dataset this size.

## 3. Train a Logistic Regression baseline

**What:** Fit a Logistic Regression model.

**Why:** Logistic Regression is fast, interpretable, and the standard first model for binary classification. It sets the **baseline score** that every more complex model later has to justify beating. (`max_iter=1000` simply gives the solver room to converge.)

In [3]:
baseline_model = LogisticRegression(max_iter=1000)
baseline_model.fit(X_train, y_train)

y_pred = baseline_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"Baseline accuracy: {accuracy:.3f}")

Baseline accuracy: 0.819


**~82% accuracy** sounds impressive — but remember from Notebook 01 that ~73% of customers don't churn. So a model that simply predicted "no churn" for everyone would already score ~73%. Our baseline is only modestly better than that naive guess, which is the warning sign we investigate properly in the next notebook.

## Section conclusion — baseline established

- Split the data into **5,634 train / 1,409 test** customers.
- Trained a **Logistic Regression baseline** scoring **~82% accuracy**.
- Flagged that this accuracy is inflated by the class imbalance — accuracy alone is not good enough.

**Next:** Notebook 06 looks past accuracy at precision, recall, and the confusion matrix to see what the baseline really does.